reliabletopic/src/main/java/com/hazelcast/jet/contrib/reliabletopic/ReliableTopicSource.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.contrib.reliabletopic;
+
+import com.hazelcast.function.FunctionEx;
+import com.hazelcast.jet.core.Processor.Context;
+import com.hazelcast.jet.pipeline.SourceBuilder;
+import com.hazelcast.jet.pipeline.StreamSource;
+import com.hazelcast.topic.Message;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayDeque;
+
+import java.util.Queue;
+
+
+
+/**
+ * Contains a method for creating a reliable topic stream sources.
+ */
+public final class ReliableTopicSource {
+
+    private ReliableTopicSource() {
+    }
+
+    /**
+     * Creates a {@link StreamSource} which reads messages from
+     * from a Hazelcast reliable topic with the specified name.
+     * @param topic       topic name to read from
+     * @param projectionFn built in mapping function of the source which can be
+     *                     used to map {@link Message} instances received
+     *                     from a reliable topic to an arbitrary type of output;
+     *                     it's included for convenience.
+     * @param <M> type of item inside a message.
+     * @param <T> type of emitted item after projection.
+     * @return {@link StreamSource}
+     */
+    public static <M, T> StreamSource<T> topicSource(
+            @Nonnull String topic,
+            @Nonnull FunctionEx<Message<M>, T> projectionFn
+    ) {
+        return SourceBuilder.timestampedStream("reliable-topic-source",
+                ctx -> new ITopicSourceContext<>(ctx, topic, projectionFn))
+                .<T>fillBufferFn(ITopicSourceContext::fillBuffer)
+                .build();
+    }
+
+    private static final class ITopicSourceContext<M, T> {
+        private static final int QUEUE_CAP = 1024;
+        private static final int MAX_FILL_MESSAGES = 128;
+
+        private final Queue<Message<M>> queue = new ArrayDeque<>(QUEUE_CAP);
+        private final FunctionEx<Message<M>, T> projectionFn;
+
+        private ITopicSourceContext(Context ctx, String topic, @Nonnull FunctionEx<Message<M>, T> projectionFn) {
+            ctx.jetInstance().<M>getReliableTopic(topic).addMessageListener(queue::add);
```

Comment:
1. `ArrayDeque` is non-concurrent, we must use a concurrent queue because the message listener adds to it in a different thread. A SPSC queue is probably sufficient, as per the specification of `MessageListener`.

2. We don't apply any upper bound on the queue. If the job is stuck or slow, the queue could grow big. Not sure how or if can we apply backpressure, needs to be investigated and/or documented.

Reply:
1) I need to use thread-safe collection here 👍 . I also talked to Ali about this but we have trouble about which thread-safe collection I should choose here. Whether I need to choose bounded one or unbounded one? The reliable topic is already using RingBuffer, which is an internally bounded collection. If I want to use bounded collection, I need to either apply backpressure or use something like blocking queue(which is against Jet's policy on blocking calls). Also, I cannot apply backpressure on this message listener logic, [here](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java#L141-L144). Since we cannot apply backpressure, it is possible to encounter problems such as OOM if we use unbounded collections as you already said.
2) I stopped using the blocking queue just because the thread was blocked. However, although I remove to use a collection having a blocking operation, I am still faced with a blocking call in this source. The thread is blocked again when no message is sent to the topic, since the messageListener that is added to the topic read from this topic in a blocking manner, [here](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/MessageRunner.java#L89-L90). I think this is a problem for us.

-> This reliable topic is using internally ring buffer maybe we can implement our own reading logic from ring buffer.